src/placement/tests/jump_map_place_obj.c

```diff
@@ -0,0 +1,207 @@
+	daos_obj_generate_id(&oid, 0, OC_RP_4G2, 0);
+	D_PRINT("\ntest initial placement when no failed shard ...\n");
+	plt_obj_place(oid, &lo_1, pl_map);
+	plt_obj_layout_check(lo_1);
+
+	/* test plt_obj_place when some/all shards failed */
+	D_PRINT("\ntest to fail all shards  and new placement ...\n");
+	for (i = 0; i < SPARE_MAX_NUM && i < lo_1->ol_nr; i++)
+		plt_fail_tgt(lo_1->ol_shards[i].po_target, &po_ver, po_map,
+				pl_debug_msg);
+	plt_obj_place(oid, &lo_2, pl_map);
+	plt_obj_layout_check(lo_2);
+	D_ASSERT(!pt_obj_layout_match(lo_1, lo_2, DOM_NR));
+	D_PRINT("spare target candidate:");
+	for (i = 0; i < SPARE_MAX_NUM && i < lo_1->ol_nr; i++) {
+		spare_tgt_candidate[i] = lo_2->ol_shards[i].po_target;
+		D_PRINT(" %d", spare_tgt_candidate[i]);
+	}
+	D_PRINT("\n");
+
+	D_PRINT("\ntest to add back all failed shards and new placement ...\n");
+	for (i = 0; i < SPARE_MAX_NUM && i < lo_1->ol_nr; i++)
+		plt_add_tgt(lo_1->ol_shards[i].po_target, &po_ver, po_map,
+				pl_debug_msg);
+	plt_obj_place(oid, &lo_3, pl_map);
+	plt_obj_layout_check(lo_3);
+	D_ASSERT(pt_obj_layout_match(lo_1, lo_3, DOM_NR));
+
+	/* test pl_obj_find_rebuild */
+	D_PRINT("\ntest pl_obj_find_rebuild to get correct spare tagets ...\n");
+	failed_tgts[0] = lo_3->ol_shards[0].po_target;
+	failed_tgts[1] = lo_3->ol_shards[1].po_target;
+	D_PRINT("failed target %d[0], %d[1]\n, expected %d[0], %d[1]\n",
+		failed_tgts[0], failed_tgts[1], spare_tgt_candidate[0],
+		spare_tgt_candidate[1]);
+	plt_spare_tgts_get(pl_uuid, oid, failed_tgts, 2, spare_tgt_ranks,
+			pl_debug_msg, shard_ids, &spare_cnt, &po_ver,
+			PL_TYPE_JUMP_MAP, SPARE_MAX_NUM, po_map, pl_map);
+	D_ASSERT(spare_cnt == 2);
+	D_ASSERT(spare_tgt_ranks[0] == spare_tgt_candidate[0]);
+	D_ASSERT(spare_tgt_ranks[1] == spare_tgt_candidate[1]);
+
+	/* fail the to-be-spare target and select correct next spare */
+	failed_tgts[0] = lo_3->ol_shards[0].po_target;
+	failed_tgts[1] = lo_3->ol_shards[1].po_target;
+	failed_tgts[2] = spare_tgt_candidate[0];
+	D_PRINT("\nfailed targets %d[1] %d %d[0], expected spare %d[1]\n",
+		failed_tgts[0], failed_tgts[1], failed_tgts[2],
+		spare_tgt_candidate[1]);
+	plt_spare_tgts_get(pl_uuid, oid, failed_tgts, 3, spare_tgt_ranks,
+			   pl_debug_msg, shard_ids, &spare_cnt, &po_ver,
+			   PL_TYPE_JUMP_MAP, SPARE_MAX_NUM, po_map, pl_map);
+	D_ASSERT(spare_cnt == 2);
+	D_ASSERT(spare_tgt_ranks[0] == spare_tgt_candidate[1]);
+
+	failed_tgts[0] = spare_tgt_candidate[0];
+	failed_tgts[1] = spare_tgt_candidate[1];
+	failed_tgts[2] = lo_3->ol_shards[3].po_target;
+	failed_tgts[3] = lo_3->ol_shards[0].po_target;
+	failed_tgts[4] = lo_3->ol_shards[1].po_target;
+	D_PRINT("\nfailed targets %d %d %d[3] %d[0] %d[1]\n",
+		failed_tgts[0], failed_tgts[1], failed_tgts[2], failed_tgts[3],
+		failed_tgts[4]);
+	plt_spare_tgts_get(pl_uuid, oid, failed_tgts, 5, spare_tgt_ranks,
+			   pl_debug_msg, shard_ids, &spare_cnt, &po_ver,
+			   PL_TYPE_JUMP_MAP, SPARE_MAX_NUM, po_map, pl_map);
```

Comment:
these codes looks very similar as original place_obj.c (current ring_map_place_obj.c), why need to be two separate UT, are they with different expected result? can that two be combined as one?

Reply:
Yes, the expected result of the tests for ring map and jump map will be different.
Because they work differently, a specific test for the ring map might not be valid for the jump map even though both are functioning correctly. 
In this case the "order" in which targets are failed makes a difference and asserts that are true for the ring map will not be true for the jump map.

Also in the future there may be additional placement maps and there is no guarantee they will function in a similar manner.